lib/os/fdtable.c

```diff
@@ -156,27 +160,64 @@ int z_alloc_fd(void *obj, const struct fd_op_vtable *vtable)
 
 #ifdef CONFIG_POSIX_API
 
-ssize_t read(int fd, void *buf, size_t sz)
+ssize_t z_impl__read(int fd, void *buf, size_t sz)
 {
 	if (_check_fd(fd) < 0) {
 		return -1;
 	}
 
 	return fdtable[fd].vtable->read(fdtable[fd].obj, buf, sz);
 }
-FUNC_ALIAS(read, _read, ssize_t);
 
-ssize_t write(int fd, const void *buf, size_t sz)
+#ifdef CONFIG_USERSPACE
+ssize_t z_vrfy__read(int fd, void *buf, size_t sz)
+{
+	if (Z_SYSCALL_MEMORY_WRITE(buf, sz)) {
+		errno = EFAULT;
+		return -1;
+	}
+
+	return z_impl__read(fd, buf, sz);
+}
+#include <syscalls/_read_mrsh.c>
+#endif /* CONFIG_USERSPACE */
+
+/* Normal C function wrapping a corresponding syscall. Required to ensure
+ * classic C linkage.
+ */
+ssize_t read(int fd, void *buf, size_t sz)
+{
+	return _read(fd, buf, sz);
+}
+
+ssize_t z_impl__write(int fd, const void *buf, size_t sz)
 {
 	if (_check_fd(fd) < 0) {
 		return -1;
 	}
 
 	return fdtable[fd].vtable->write(fdtable[fd].obj, buf, sz);
 }
-FUNC_ALIAS(write, _write, ssize_t);
 
-int close(int fd)
+#ifdef CONFIG_USERSPACE
+ssize_t z_vrfy__write(int fd, const void *buf, size_t sz)
+{
+	Z_OOPS(Z_SYSCALL_MEMORY_READ(buf, sz));
+
+	return z_impl__write(fd, buf, sz);
+}
+#include <syscalls/_write_mrsh.c>
+#endif /* CONFIG_USERSPACE */
+
+/* Normal C function wrapping a corresponding syscall. Required to ensure
+ * classic C linkage.
+ */
+ssize_t write(int fd, const void *buf, size_t sz)
+{
+	return _write(fd, buf, sz);
```

Comment:
Getting "undefined reference to `_write`" and "undefined reference to `_read`" in libc_nano.a . It doesn't jump out at me immediately why that might be the case.
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/73259/5/console

Reply:
... But it should explain why this spent so much time in the backlog - dealing with all these naming (ab)uses across different targets is a bit boring (and I'm not particularly interested in kernel/userspace split ;-) ). Anyway, fixing it.